File: SuperpositionGame/src/main/scala/superposition/Complex.java
package superposition;

public class Complex {

    public final double real, im;

    public Complex(double real, double imaginary) {
        this.real = real;
        this.im = imaginary;
    }

    public Complex(double real) {
        this.real = real;
        this.im = 0;
    }

    public Complex add(Complex o) {
        return new Complex(real + o.real, im + o.im);
    }

    public Complex div(Complex o) {
        var m = o.magnitudeSquared();
        return new Complex((real * o.real + im * o.im) / m, (im * o.real - real * o.im) / m);
    }

    public double magnitude() {
        return Math.sqrt(magnitudeSquared());
    }

    public double magnitudeSquared() {
        return real * real + im * im;
    }

    public Complex mul(Complex o) {
        return new Complex(real * o.real - im * o.im, real * o.im + im * o.real);
    }

    public double phase() {
        return Math.atan2(im, real);
    }

    public static Complex polar(double r, double theta) {
        return new Complex(r * Math.cos(theta), r * Math.sin(theta));
    }

    public Complex sub(Complex o) {
        return new Complex(real - o.real, im - o.im);
    }
}

File: SuperpositionGame/src/main/scala/superposition/Complex.scala
package superposition

import scala.math.{atan2, cos, sin, sqrt}

private object Complex {
  def polar(r: Double, theta: Double) = Complex(r * cos(theta), r * sin(theta))
}

private case class Complex(real: Double, imag: Double = 0.0) {
  def +(that: Complex): Complex = Complex(real + that.real, imag + that.imag)

  def -(that: Complex): Complex = Complex(real - that.real, imag - that.imag)

  def *(that: Complex): Complex = Complex(
    real * that.real - imag * that.imag,
    real * that.imag + imag * that.real
  )

  def /(that: Complex): Complex = Complex(
    (real * that.real + imag * that.imag) / that.magnitudeSquared,
    (imag * that.real - real * that.imag) / that.magnitudeSquared
  )

  def magnitude: Double = sqrt(magnitudeSquared)

  def magnitudeSquared: Double = real * real + imag * imag

  def phase: Double = atan2(imag, real)
}
